api/src/main/java/org/apache/iceberg/stats/StatsUtil.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.stats;
+
+import static org.apache.iceberg.types.Types.NestedField.optional;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.types.Types;
+
+public class StatsUtil {
+  private static final int NUM_STATS_PER_COLUMN = 200;
+  static final int RESERVED_FIELD_IDS = 200;
+  static final int DATA_SPACE_FIELD_ID_START = 10_000;
+  static final int METADATA_SPACE_FIELD_ID_START = 1_417_000_000;
+  static final int RESERVED_FIELD_IDS_START = Integer.MAX_VALUE - RESERVED_FIELD_IDS;
+
+  private StatsUtil() {}
+
+  public static int statsFieldIdFor(int fieldId) {
+    int idSpaceStart = DATA_SPACE_FIELD_ID_START;
+    int id = fieldId;
+    if (fieldId >= RESERVED_FIELD_IDS_START) {
+      // this is a reserved field ID, which uses a different calculation
+      idSpaceStart = METADATA_SPACE_FIELD_ID_START;
+      id = RESERVED_FIELD_IDS - (Integer.MAX_VALUE - fieldId);
+    }
+
+    long finalId = idSpaceStart + NUM_STATS_PER_COLUMN * (long) id;
+    if (finalId < 0
+        || finalId > RESERVED_FIELD_IDS_START
+        || (finalId >= METADATA_SPACE_FIELD_ID_START
+            && idSpaceStart != METADATA_SPACE_FIELD_ID_START)) {
+      // ID overflows
+      return -1;
+    }
+
+    return (int) finalId;
+  }
+
+  public static int fieldIdFor(int statsFieldId) {
+    if (statsFieldId < 0 || statsFieldId % NUM_STATS_PER_COLUMN != 0) {
```

Comment:
This one isn't as complicated as its inverse but I still would prefer to separate out the math into specific functions, that way we don't need the comment either.